lib/Sema/CSDiag.cpp

```diff
@@ -4730,13 +4733,115 @@ static bool diagnoseSingleCandidateFailures(CalleeCandidateInfo &CCI,
     }
 
     void missingArgument(unsigned missingParamIdx) override {
-      Identifier name = Parameters[missingParamIdx].Label;
-      auto loc = ArgExpr->getStartLoc();
+      auto &param = Parameters[missingParamIdx];
+      Identifier name = param.Label;
+
+      // Search insertion index.
+      unsigned argIdx = 0;
+      for (int Idx = missingParamIdx - 1; Idx >= 0; --Idx) {
+        if (Bindings[Idx].empty()) continue;
+        argIdx = Bindings[Idx].back() + 1;
+        break;
+      }
+
+      unsigned insertableEndIdx = Arguments.size();
+      if (CandidateInfo.hasTrailingClosure)
+        insertableEndIdx -= 1;
+
+      // Build argument string for fix-it.
+      SmallString<32> insertBuf;
+      llvm::raw_svector_ostream insertText(insertBuf);
+
+      if (argIdx != 0)
+        insertText << ", ";
+      if (!name.empty())
+        insertText << name.str() << ": ";
+      Type Ty = param.Ty;
+      // Explode inout type.
+      if (auto IOT = param.Ty->getAs<InOutType>()) {
+        insertText << "&";
+        Ty = IOT->getObjectType();
+      }
+      // @autoclosure; the type should be the result type.
+      if (auto FT = param.Ty->getAs<AnyFunctionType>())
+        if (FT->isAutoClosure())
+          Ty = FT->getResult();
+      insertText << "<#T##" << Ty << "#>";
```

Comment:
IIRC, the syntax for a typed completion token is `<#name##type#>`; there's nothing special about "T". We should then show the internal name of the parameter, like code completion does. @benlangmuir, thoughts?